### ISSUE TITLE
docs: recommendations on how to handle "dependency injection"

### ISSUE THREAD

**Issue by Nick-Lucas** [trpc/trpc#3958]
### Area of Improvement

A lot of new users are unsure how to get their services, orms, api clients, etc, into their procedures. There are only 2-3 vanilla tRPC options available and they're quite straight-forward to explain, so we should write this down

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR implementing the suggested changes!

**Comment by radeon8** [comment]
It is nice 👍

**Comment by c-ehrlich** [comment]
The first sentence of the current context page is "Your context holds data that all of your tRPC procedures will have access to, and is a great place to put things like database connections or authentication information." The page also uses prisma as an example in one of the code blocks.

I think this is less of a question of adding more information, and more of a bigger issue of the docs as a whole being structured in an unintuitive way currently.

**Comment by Nick-Lucas** [comment]
Agreed, my approach is we can do a lot by starting with needs and patterns rather than APIs. You shouldn't have to read the docs front to back

**Comment by ben-xD** [comment]
I'm curious how people are injecting services into their context

> Vanilla recommendations are:
>     - singletons
>     - createContext
>     - middlewares with context swapping

- Why singletons are bad: makes testing harder or changing code harder. I wouldn't call singletons a way of "dependency injection". IIRC @KATT is not a fan of unit testing.
- Why not createContext: We can't (or shouldn't) really instantiate services inside createContext, since they'd be created **on each request**. 
- Why not middlewares with context swapping: I don't see how this solves the problem. That's just for changing the context type.

When trying to actually **use dependency injection**, I find the types provided by tRPC to be hard to work with. I'm not sure how to type `CreateContext` so that I can return it from a function. For example, what should `CreateContext` be?

```
export type Services = {
  orderService: OrderService;
  productService: ProductService;
  inventoryService: InventoryService;
};

export type CreateContext = NodeHTTPHandlerOptions<
AnyRouter,
NodeHTTPRequest,
NodeHTTPResponse
>;
export type Context = inferAsyncReturnType<CreateContext>;

// Create a createContext function that will contain all services needed in the context, 
// so that services are not created inside createContext.
export const createCreateContext = (services: Services): CreateContext => {
  
  // The context is created on every request. Do not instantiate long term objects inside createContext.
  // docs: https://trpc.io/docs/context
  const createContext = (
    _options: (CreateExpressContextOptions | CreateWSSContextFnOptions),
  ): Services => {
    // Authentication can be added here
    return { ...services };
  };

  return createContext;
}
```

**Comment by c-ehrlich** [comment]
What's wrong with something like:
```ts
import { someServiceSingleton } from "../util/some-service";

export const createTRPCContext = () => {
  return { someService: someServiceSingleton };
};
```
this will not be recreated on every request in a serverfull environment

alternatively or additionally:
```ts
const swapSomeService = middleware(async ({ ctx, next }) => {
  return next({
    ctx: {
      someService: somethingElse,
    },
  });
});
```

**Comment by ben-xD** [comment]
The issue for me is that `someServiceSingleton` and `somethingElse` is a singleton 🤓. I'd like to minimise the usage of global variables. What if I don't want to use singleton, or I want to configure that service class (e.g. by specifying some constructor arguments).

I want to minimise the usage of singletons to avoid spaghetti. Forcing constructor dependency injection is useful - the hierarchy of classes becomes very clear.

**Comment by ben-xD** [comment]
> There's nothing inherently wrong with singletons, they're a fundamental part of work in javascript.

The discussion on using singleton is a long and windy one - probably not worth going too deeply about - down to personal preference to an extent. I do think it's fair that some developers would want to avoid singleton in Javascript, and that it should be supported.

My main issue is that it's difficult (impossible?) to avoid singletons in this case. I've seen singletons and global variables go wrong quite a few times. It enables importing whatever you need, anywhere. Anyone could import those service global variables from anywhere instead of using Context. Also, these services might have dependencies on other global variables - eventually it becomes a mess. Just importing the file that instantiates a global variable can cause side effects.

> JS testing frameworks typically have very powerful mocking/stubbing tools, so you can replace a singleton's implementation with a stub during testing

Yea, I'm currently testing my routers in tRPC: e.g. using vitest's [`mock` function](https://vitest.dev/api/vi.html#vi-mock). That's why I did not bring up tests. :)

**Comment by Nick-Lucas** [comment]
> Anyone could import those service global variables from anywhere instead of using Context

It's true, but anyone could import and instantiate a service which should have been retrieved from DI too. It's sort of on us as developers to stay on top of misuse. eslint can also be set up to ban imports from specific directories if you like!

**Comment by DirectStack** [comment]
It is my opinion just as a normal developer.
My projects are small, sometimes medium. tRPC and React support better DX and make my life easy.
If I need DI, and the project maybe very large, why not Nestjs with OOP?
In my own case, DI costs more than business. And switch from Nestjs to tRPC.
If there is a mistake, pls point out.

**Comment by libasoles** [comment]
Hi. My take on this:
 _tRPC could recommend a third-party DI solution instead of developing one_.

I know tRPC provides a context, and that's great to access stuff in the procedures. But what about other objects needing DI as well? E.g:

I like to separate concerns in the backend like `procedures/controllers > services > gateways/repositories`

So **UserService** would benefit from receiving the **usersRepository**.

```javascript
const User = require('./User');

function UsersService(usersRepository) {
  async function getUsers() {
    return usersRepository.findAll();
  }
  
  async function addUser(userData) {
    const user = new User(userData);
  
    return usersRepository.addUser(user);
  }
  
  return {
    getUsers,
    addUser
  };
}

module.exports = UsersService
```

Getting back to my first point, I don't think that tRPC context "_is a great place to put things like database connections_". Because from my architecture perspective, procedures shouldn't connect to databases at all. That's the responsibility of repositories.

Another problem is the complexity you were discussing. Setting up the factories and singletons using different configuration sources. I've worked with a couple of DI contexts and the best experience I had was using annotations with **Spring**. And I've seen the annotations approach in **Nest** as well, but I don't know if the DI solution they have is a module that could be used outside Nest.

I do know after reading this article (https://tsh.io/blog/dependency-injection-in-node-js/) that there are a few libraries for DI in node:

- awilix
- inversify
- typeDI

I'll make a PoC with one of them. My goal is avoiding to setup the DI container as much as possible. I kinda like magic, at least for most common scenarios (where no config file and stuff like that are needed).

**Comment by Nick-Lucas** [comment]
Apologies this has taken a while to really get back to. I know there's a lot of interest in solutions and patterns, we've just been focused on bigger structural issues in the docs first.

@libasoles If you're looking to integrate a DI solution into tRPC something like this might be a good starting point: https://stackblitz.com/edit/github-ewmssd?file=src/InjectorMiddleware.ts 

> I don't think that tRPC context "is a great place to put things like database connections". Because from my architecture perspective, procedures shouldn't connect to databases at all. That's the responsibility of repositories.

The phrasing could be a bit better, but it's a bit of a nit really. You can put your ORM in there too and make it available to procedures trivially. Service classes which consume other service classes do start to require DI though and we don't have a 1st party solution for that beyond the stackblitz prototype I shared. You can also use middlewares to inject just to a subset of procedures using [Context Extension](https://trpc.io/docs/server/middlewares#context-extension)

**Comment by akashdevcc** [comment]
@Nick-Lucas @ben-xD @libasoles 

Here is my effort on achieving "dependency injection" in tRPC using nestjs.
https://github.com/akashdevcc/zero-stack
Everything from the creating the `AppRouter` and `createContext` is injected using nestjs.

Custom decorators for tRPC routers and procedures can also be implemented later on, to provide more abstraction to this.

But there is a caveat, I was able to make it work up to the server app part, where I am able to successfully perform the tRPC operations on the tRPC playground. 

But, I am getting TypeScript errors on the client apps, for which I have created a new issue https://github.com/trpc/trpc/issues/5614 here.

Feel free to have a look if you know how this issue can be resolved.

**Comment by sbue** [comment]
@akashdevcc This is extremely timely for me. Well done :clap:

**Comment by akashdevcc** [comment]
@sbue Please check if you are able to resolve the issue of property collision typescript error

**Comment by ben-xD** [comment]
I haven't had time to dig into your project @akashdevcc. However, I have solved this problem a few times. My current approach is a `createTrpcCreateContext` function. It's pretty simple, and haven't had any issues.

See https://github.com/ben-xD/talkdash/blob/ec205bcbcde1dcb42ac5b2896ae47ed1c8677810/backend/src/trpc/fastifyTrpcRoutes.ts#L21
and https://github.com/ben-xD/talkdash/blob/ec205bcbcde1dcb42ac5b2896ae47ed1c8677810/backend/src/trpc/trpcContext.ts#L11-L49

**Comment by akashdevcc** [comment]
@ben-xD Your approach is good. It is suitable for small-scale projects but for a large-scale projects where you need to inject hundreds of dependencies coming from serveral modules, you are going to need a matured framework eventually. NestJS solves that.

Also, the issue I was getting has now been resolved. **And, "tRPC + NestJS" integration is now successfully done end-to-end, along with working example of expo and nextjs apps.**

Anyone, who'd like to have "dependency injection" in tRPC, have a look at this: https://github.com/akashdevcc/zero-stack
And let me know your thoughts/feedback/suggestions.
